File: frontend/src/api/tracked.ts
import { api } from './client'

export interface TrackedItem {
  id: number
  marketplace: string
  sku: string
  name: string
  url: string
  image_url: string | null
  initial_price: number
  current_price: number
  last_checked_at: string
  alert_threshold_pct: number
  is_active: boolean
  created_at: string
}

export interface PricePoint {
  price: number
  captured_at: string
}

export async function getTracked(): Promise<TrackedItem[]> {
  return api.get<TrackedItem[]>('/tracked')
}

export async function addTracked(data: {
  marketplace: string
  sku: string
  name: string
  url: string
  image_url?: string | null
  initial_price: number
}): Promise<TrackedItem> {
  return api.post<TrackedItem>('/tracked', data)
}

export async function removeTracked(id: number): Promise<void> {
  await api.delete(`/tracked/${id}`)
}

export async function getPriceHistory(id: number): Promise<PricePoint[]> {
  return api.get<PricePoint[]>(`/tracked/${id}/history`)
}

File: backend/app/models/tracked.py
from datetime import datetime

from sqlalchemy import BigInteger, Boolean, ForeignKey, Integer, String, Text
from sqlalchemy.orm import Mapped, mapped_column
from sqlalchemy.sql import func

from app.db import Base
from app.models._types import TimestampTZ


class TrackedItem(Base):
    __tablename__ = "tracked_items"

    id: Mapped[int] = mapped_column(BigInteger, primary_key=True, autoincrement=True)
    user_id: Mapped[int] = mapped_column(BigInteger, ForeignKey("users.id", ondelete="CASCADE"), nullable=False)
    created_at: Mapped[datetime] = mapped_column(TimestampTZ, server_default=func.now())

    marketplace: Mapped[str] = mapped_column(String(32), nullable=False)
    sku: Mapped[str] = mapped_column(Text, nullable=False)
    name: Mapped[str] = mapped_column(Text, nullable=False)
    url: Mapped[str] = mapped_column(Text, nullable=False)
    image_url: Mapped[str | None] = mapped_column(Text)

    initial_price: Mapped[int] = mapped_column(Integer, nullable=False)
    current_price: Mapped[int] = mapped_column(Integer, nullable=False)
    last_checked_at: Mapped[datetime] = mapped_column(TimestampTZ, server_default=func.now())

    alert_threshold_pct: Mapped[int] = mapped_column(Integer, default=5)
    last_alert_sent_at: Mapped[datetime | None] = mapped_column(TimestampTZ)

    is_active: Mapped[bool] = mapped_column(Boolean, default=True)


class PriceHistory(Base):
    __tablename__ = "price_history"

    id: Mapped[int] = mapped_column(BigInteger, primary_key=True, autoincrement=True)
    tracked_item_id: Mapped[int] = mapped_column(
        BigInteger, ForeignKey("tracked_items.id", ondelete="CASCADE"), nullable=False
    )
    price: Mapped[int] = mapped_column(Integer, nullable=False)
    captured_at: Mapped[datetime] = mapped_column(TimestampTZ, server_default=func.now())
